### PR TITLE
Solid Milestone initialization fix

### PR DESCRIPTION
**Pull request by Desolatora**
### Bug
LedgerValidator initializes wrong consistent milestone after restart when node is not fully synced.

### Symptoms
When node is synced to solid milestone X and latest milestone is Y, after restart solid milestone is incorrectly initialized to Z (usually very close to Y).

### Cause
Method LedgerValidator.buildSnapshot() is responsible for building the most recent consistent milestone. It uses MilestoneViewModel.firstWithSnapshot() and MilestoneViewModel.nextWithSnapshot() to iterate on every milestone with snapshot then gets the state differences and patches them to the current snapshot. These methods deterministically rely on StateDiffViewModel.exists() which internally uses tangle.maybeHas(). The problem is that tangle.maybeHas() can only guarantee when entry is not present, but does not guarantee if entry is present. This leads to MilestoneViewModel.firstWithSnapshot() and MilestoneViewModel.nextWithSnapshot() sometimes returning milestone without snapshot or more precise milestone without StateDiff. This incorrect behaviour is also supported by the fact that StateDiffViewModel.load() returns an object even if StateDiff is not present. This leads to latestSnapshot being patched with no state differences and incorrect milestone index.

### Problems
When node is initialized with wrong solid milestone this may lead to solid milestone tracker appearing frozen while it is trying to solidify next milestone or compute state diff with both operations involving huge number of transactions. Sometimes wrong solid milestone index may lead to "Transaction resolves to incorrect ledger balance" errors (bug: #460).

### Proposed fix
Method StateDiffViewModel.isEmpty() is introduced to check if current state difference is empty. LedgerValidator.buildSnapshot() now iterates on every candidate milestone and patches latestSnapshot only with nonempty state differences. This guarantees that initialized consistent milestone is solid.

### CODE REVIEW

File: src/main/java/com/iota/iri/LedgerValidator.java [hunk 1]
```diff
@@ -221,20 +223,34 @@ public static boolean isApproved(Snapshot snapshot, Hash hash) {
      */
     private MilestoneViewModel buildSnapshot(Snapshot latestSnapshot) throws Exception {
         MilestoneViewModel consistentMilestone = null;
+
         synchronized (latestSnapshot.snapshotSyncObject) {
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/LedgerValidator.java]: Use locks. See updated code

File: src/main/java/com/iota/iri/controllers/StateDiffViewModel.java [hunk 2]
```diff
@@ -24,7 +24,7 @@ public StateDiffViewModel(final Map<Hash, Long> state, final Hash hash) {
         this.stateDiff.state = state;
     }
 
-    public static boolean exists(Tangle tangle, Hash hash) throws Exception {
+    public static boolean maybeExists(Tangle tangle, Hash hash) throws Exception {
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/controllers/StateDiffViewModel.java]: Like

**Comment by Desolatora** [comment]
Rebased with new code.

**Comment by GalRogozinski** [comment]
Hey @Desolatora,
It seems like the rebase didn't roll out as smoothly as we like.
If you click on the "commits" tab or the "file changes" tab you will immediately see what I mean.

Git can be annoying sometimes, but we need to get it fixed.
If you find it too hard to fix this PR, you can close it and issue a new PR with only a link to this one in the description.

**Comment by Desolatora** [comment]
Rebase fixed.

**Comment by Desolatora** [comment]
It should be fixed now. If it isn't i will create a new PR.

**Comment by Desolatora** [comment]
buildSnapshot() will now stop searching for candidates if inconsistent milestone is encountered.
Also i removed the log.isDebugEnabled() part and actually changed the log level to INFO. I think it is essencial for iri to report milestone initialization, because this is fairly slow process and it may appear that iri is stuck while actually it is properly initializing.

**Pull request merged**